### ISSUE TITLE
token renewal is not thread-safe

### ISSUE THREAD

**Issue by majewsky** [gophercloud/gophercloud#310]
I can't see anything in the default token reauth functions ([`v2auth`](https://github.com/gophercloud/gophercloud/blob/9a5595b8ffe429439c63781cdc312c254e4ad990/openstack/client.go#L93) and [`v3auth`](https://github.com/gophercloud/gophercloud/blob/9a5595b8ffe429439c63781cdc312c254e4ad990/openstack/client.go#L144)) that protects against multiple threads trying to renew a single client's token at the same time.

I worked around this in my client code by implementing a custom reauth function that [locks a mutex](https://github.com/sapcc/limes/blob/master/pkg/limes/realdriver.go#L81-L82), and this seems to mostly work, but it's not perfect: There can still be duplicate token renewals when multiple threads enter the reauth function at the same time. (When one reauth function call is done, the mutex is unlocked, and the next thread will take the lock and reauth again immediately.)

In any way, it would be nice if gophercloud offers thread-safety by default here, so that ProviderClient instances can be shared between threads without any weird surprises.

**Comment by armstrongli** [comment]
We got the same issue in our production when there're multiple go routines sending requests at the same time. And we have a serious problem that if one of the openstack service is unstable, this client sends infinite re-auth to keystone.

**Comment by jrperritt** [comment]
I'll commit to thinking about this, but not necessarily to doing anything about it. Gophercloud's view on this is that concurrency should be handled in the application.

> And we have a serious problem that if one of the openstack service is unstable, this client sends infinite re-auth to keystone.

@armstrongli A gist of some logs would be helpful. You've claimed that Gophercloud is causing your issues (and it may well be), but you haven't provided any proof.

**Comment by armstrongli** [comment]
@jrperritt There's one case that there're 100(e.g.) go routines do requests, and token gets expired, and go routines start doing reauth. the shared client token is updated by different go routines at different time without any thread safe mechanism. 
Here is one case which causing issues: routine A starts sending service request after `reauth`; routine B start reauth as the previous call failed with 401, and sets the token to be `""`. routine A fails to send service request and gets 401 error.
routine B could encounter the same issue as routine A.

**Comment by jrperritt** [comment]
@armstrongli Yes, I understand what the code does; I asked for the logs that show what you're describing. Either way, as I mentioned above, all that synchronization can and should be handled in your application.

**Comment by armstrongli** [comment]
The logs show that there're a lot of reauth when the token gets expires. There should be only one reauth.

```
...
Reauthing
Reauthing
Reauthing
Reauthing
Reauthing
Reauthing
Reauthing
Reauthing
...
```

It is supposed to have only one time reauth.

**Comment by tatsuhiro-t** [comment]
Currently, when our system need to access OpenStack, it create new ProviderClient using openstack.AuthenticatedClient, and do its job.  But it is not efficient if we have hundreds of concurrent jobs doing those stuff.  Each time, we create new ProviderClient, keystone creates new token.  Those tokens pile up, and pressures database.
We are now considering to share ProviderClient between goroutine so that as long as a token is valid, we keep using it, and only renew it when it is expired.  And we hit this issue.
@jrperritt Could you show how to deal with this situation by application code?  If I understand correctly, you seem to have an idea to achieve this without changing gophercloud at all.  Sorry if I misunderstand that.

**Comment by jrperritt** [comment]
@tatsuhiro-t Good question. In general, things that an application wants to control around the HTTP client can be customized similar to how it's done in the [FAQ](https://github.com/gophercloud/gophercloud/blob/master/FAQ.md#implementing-default-logging-and-re-authentication-attempts ).

However, after considering this particular use-case of concurrency, I've decided that this is probably something Gophercloud should allow for. The library puts the token in an object that's expected to be create-once-write-many, so I'll send a PR for this.

**Comment by tatsuhiro-t** [comment]
@jrperritt Thank you.  I'm looking forward to the PR.

**Comment by jrperritt** [comment]
closed by #636
